simulator/infra/replacement/cache_replacement.cpp

```diff
@@ -68,10 +69,131 @@ std::size_t LRUCacheInfo::update()
     return lru_elem;
 }
 
+//////////////////////////////////////////////////////////////////
+
+struct node
+{
+    node(int new_value = 0, std::size_t new_pointer = 0)
+    : value( new_value), pointer( new_pointer) {}
+
+    int value;
+    std::size_t pointer;
+};
+
+bool operator == (const node &lhs, const node &rhs) { return lhs.value == rhs.value; }
+bool operator< (const node &lhs, const node &rhs) { return lhs.value < rhs.value; }
+
+class Pseudo_LRUCacheInfo : public CacheReplacementInterface
+{
+    public:
+        explicit Pseudo_LRUCacheInfo( std::size_t ways);
+        void touch( std::size_t way) override;
+        void set_to_erase( std::size_t ) override;
+        std::size_t update() override;
+        std::size_t get_ways() const override { return ways; }
+
+    private:
+        enum flags { Left = 0, Right = 1};
+
+        core::tree<node> lru_tree;
+        void construct_tree( core::tree<node>::iterator node_it, std::size_t max_depth);
+        void construct_leaf_layer();
+        std::size_t calculate_depth() const;
+        std::size_t which_sibling( core::tree<node>::iterator node_it);
+        static int reverse_flag( int flag);
+
+        const std::size_t ways;
+        int leaf_iterator = 0;
+        int node_iterator = INT8_MIN; //make shure it is less than 0
+};
+
+Pseudo_LRUCacheInfo::Pseudo_LRUCacheInfo( std::size_t ways)
+    : ways( ways)
+{
+    lru_tree.data( node(node_iterator));
+    construct_tree( lru_tree.get_tree_iterator(), calculate_depth());
+}
+
+std::size_t Pseudo_LRUCacheInfo::calculate_depth() const
+{
+    std::size_t i = 0;
+        for ( std::size_t numerator = ways; numerator != 1; numerator = numerator/2)
+        {
+            if (numerator%2 != 0)
+                throw CacheReplacementException("Number of ways must be the power of 2!");
+            i++;
+        }
+    return i;
+}
+
+void Pseudo_LRUCacheInfo::construct_tree( core::tree<node>::iterator node_it, std::size_t max_depth)
+{
+    if ( node_it.level() < max_depth)
+    {
+        if ( node_it.level() != max_depth - 1)
+        {
+            auto left_it = node_it.insert( node( node_iterator++, Left));
+            auto right_it = node_it.insert( node( node_iterator++, Left));
+            construct_tree( left_it, max_depth);
+            construct_tree( right_it, max_depth);
+        }
+        else
+        {
+            node_it.insert( node(leaf_iterator++));
+            node_it.insert( node(leaf_iterator++));
+        }
+    }
+}
+
+int Pseudo_LRUCacheInfo::reverse_flag( int flag) //more readable than multiplying on -1
```

Comment:
The input parameter should be of `enum flag` type. Additionally, we usually name types With Capital Letters.